libbeat/outputs/shipper/shipper.go

```diff
@@ -153,37 +172,67 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
 
 	ctx, cancel := context.WithTimeout(ctx, c.timeout)
 	defer cancel()
-	resp, err := c.client.PublishEvents(ctx, &messages.PublishRequest{
+	publishReply, err := c.client.PublishEvents(ctx, &messages.PublishRequest{
+		Uuid:   c.serverID,
 		Events: convertedEvents,
 	})
 
-	if status.Code(err) != codes.OK || resp == nil {
+	if status.Code(err) != codes.OK || publishReply == nil {
 		batch.Cancelled()         // does not decrease the TTL
 		st.Cancelled(len(events)) // we cancel the whole batch not just non-dropped events
 		return fmt.Errorf("failed to publish the batch to the shipper, none of the %d events were accepted: %w", len(convertedEvents), err)
 	}
 
 	// with a correct server implementation should never happen, this error is not recoverable
-	if int(resp.AcceptedCount) > len(nonDroppedEvents) {
+	if int(publishReply.AcceptedCount) > len(nonDroppedEvents) {
 		return fmt.Errorf(
 			"server returned unexpected results, expected maximum accepted items %d, got %d",
 			len(nonDroppedEvents),
-			resp.AcceptedCount,
+			publishReply.AcceptedCount,
 		)
 	}
 
-	// the server is supposed to retain the order of the initial events in the response
-	// judging by the size of the result list we can determine what part of the initial
-	// list was accepted and we can send the rest of the list for a retry
-	retries := nonDroppedEvents[resp.AcceptedCount:]
+	if publishReply.PersistedIndex < publishReply.AcceptedIndex {
+		ackClient, err := c.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+			PollingInterval: durationpb.New(c.config.AckPollingInterval),
```

Comment:
We need to be careful here as the polling interval will effectively set a limit on the maximum throughput per worker. 

Per https://github.com/elastic/elastic-agent-shipper/pull/95#discussion_r947897465 we will probably always need to make this RPC call, and it is likely that we will need at least one polling interval to elapse to see the update we want. The minimal delay between receiving the `PublishResponse` and making this RPC may not be enough for the value of PersistentIndex to update. In the worst case we always incur at least one `AckPollingInterval` duration before being acknowledged.

If we keep the current `5 ms` polling interval then if we were to always incur that delay for every batch then we could publish at most 1000 ms / 5 ms = 200 batches / sec. 

With the default [max_bulk_size](https://www.elastic.co/guide/en/beats/filebeat/current/elasticsearch-output.html#_bulk_max_size) of 50 events this would limit the default throughput per worker to 200 * 50 = 10000 events/second. That doesn't feel unreasonable, but assuming my calculation is correct here we should document this type of calculation as the approximate throughput limit per shipper worker.

Reply:
I'll add this to the config documentation.

Reply:
I added the calculations in https://github.com/elastic/beats/pull/32708/commits/7940a0d519b04595d0a230af35cb83ec01e4ea90

Reply:
I'm not sure this is the right way to approach it. There's no reason to wait for full acknowledgment after every batch. Doing that would mean that no matter how big we make the shipper queue, only 1 batch worth of it would ever be filled, because the input won't advance until everything in it is sent.

But accumulating pending events is what the shipper queue is for -- if a batch has been published, we can immediately move on to the next batch and publish that, even if it isn't persisted yet. In that case the client would need to keep a list of outstanding batches, and if the persisted index is being removed from the publish response then it also needs a separate goroutine listening to persisted index changes and acknowledging batches as it goes.

Reply:
(When I said in the meeting that this input needs to do things that most inputs won't, it was this: most clients can just call publish continuously and check the persisted index passively if at all. Beats is a special case because it _needs_ to advance the persisted index eventually or it gets stuck, because its own queue fills up. For inputs like filestream or kafka where the data source is relatively stable, this is redundant -- there's no particular maximum distance between the accepted and persisted indices, because there's no need for the client to keep a copy of the data. It's only because this adapter needs to handle arbitrary data from any beats queue that this case is more finicky.)

Reply:
> I'm not sure this is the right way to approach it. There's no reason to wait for full acknowledgment after every batch. Doing that would mean that no matter how big we make the shipper queue, only 1 batch worth of it would ever be filled, because the input won't advance until everything in it is sent.
>
> But accumulating pending events is what the shipper queue is for -- if a batch has been published, we can immediately move on to the next batch and publish that, even if it isn't persisted yet

This is an excellent point. We can and should take the position that the shipper client's job here is to offload data to the shipper as quickly as it can to minimize or eliminate the need to buffer anything in the Beat itself.

I'm not familiar enough with the rest of the beats pipeline yet to know how hard it would be to deal with asynchronously observing the persisted index and acknowledging batches. I can definitely see why it was done this way first as its the most natural way to satisfy the `Publish` interface.

Reply:
I think it's a matter of returning immediately from the publish call (without updating the persisted index or acking / cancelling the batch) -- then save the outstanding batches in `shipper` and run a `PersistedIndex` listener in parallel, and when an update comes in _then_ call `batch.ACK()` on the batches that are covered. That way the beats pipeline will keep sending us batches after we finish the publish RPC.

However, I don't see a way around buffering things in the beat -- this approach will duplicate every pending event in memory (once in the beats queue and once in the shipper queue) because the beats queue can't safely remove an event until it's persisted, and that doesn't happen until the shipper is also done with it.

Reply:
Actually, I do see a way around buffering things in the beat, but it'll take some fiddling with the pipeline: we could give this specific output a special queue configuration so its events are freed when they are published, but the ACKs aren't sent to the input producer until they're persisted. It will take some plumbing but probably worth it to avoid a 2x memory use.

But that is independent of this PR -- here, the ideal behavior is still to send things as fast as you can and acknowledge outstanding batches when their index is persisted.

Reply:
> Actually, I do see a way around buffering things in the beat, but it'll take some fiddling with the pipeline: we could give this specific output a special queue configuration so its events are freed when they are published, but the ACKs aren't sent to the input producer until they're persisted. It will take some plumbing but probably worth it to avoid a 2x memory use.

Excellent, can you create an issue for us to do this as a follow up? I definitely want to avoid a 2x increase in memory usage with the shipper.

Reply:
https://github.com/elastic/elastic-agent-shipper/issues/97

Reply:
Thanks! I think my preference right now is to keep this implementation as is and come back to it with the changes to https://github.com/elastic/elastic-agent-shipper/issues/97.

I think we can have an experimental shipper with this client implementation. Considering where we are in the release cycle I'd rather Denis focus on getting some integration tests written to validate the system and find any problems we don't know about yet.

Reply:
> I think it's a matter of returning immediately from the publish call (without updating the persisted index or acking / cancelling the batch) -- then save the outstanding batches in shipper and run a PersistedIndex listener in parallel, and when an update comes in then call batch.ACK() on the batches that are covered. That way the beats pipeline will keep sending us batches after we finish the publish RPC.

@faec this is exactly the approach I took first but it got really complicated very quick  and I remembered we said that the client must be as simple as possible for other developers to implement. So, that's why I ended up with this blocking implementation instead.

Doing this async would make us to do at least the following:

1. Define this new type
```go
 type pendingBatch struct {                                                                                                                                                                                                                                                   
   batch            publisher.Batch
   serverID         string
   acceptedCount    uint32
   acceptedIndex    uint64
   nonDroppedEvents []publisher.Event
 }
```
2. Once a batch is accepted on the server we send a `pendingBatch` to `chan pendingBatch` in the output's `Publish` function
3. Listen to changes of `PersistedIndex` and `UUID` using the `PersistedIndex` endpoint in a background go routine
4. Listen to a buffered `chan pendingBatch` in a background go routine
5. On each `pendingBatch` from `chan pendingBatch` we block until a signal from 3. (e.g. using `sync.Cond`) when `PersistedIndex` >= `pendingBatch.AcceptedIndex` with all the synchronisation difficulties involved

Also, there are edge cases to handle:
1. When not entire batch was accepted and we must retry the rest
2. When multiple pending batches should be acknowledged from a single `PersistedIndex` update
3. When the server UUID changes and we need to cancel all pending batches sent to this server

and perhaps many more that I didn't hit yet.

So, if our intention to keep the client simple it has to be synchronous or we stop pursuing this goal and make it complex but high-performant.

Reply:
Well, simplicity is nice but I think limiting the entire shipper queue to at most one batch is enough of a constraint that it's worth tracking some metadata.

However, I think it can be at least somewhat simpler than your suggestion. I would do something like:

(1) add a list of pending batches to the shipper struct

```go
type shipper struct {
	...
	pending []struct {
		batch publisher.Batch
		index uint64
	}
	pendingLock sync.Mutex
}
```

(2) modify `shipper.Publish` so it sends _all_ events in the batch (making multiple upstream publish calls if necessary), and saves the batch as well as the _final_ accepted index in `shipper.pending`. [This is more in keeping with the intent of the API anyway: currently the shipper only blocks until at least one event is sent, but that's just because the queue is full, not because some of the events really "failed." So marking these events as failed and needing retry in the beats output doesn't fit the usual semantics -- we don't want to be potentially dropping lots of events over "repeated failures" when all that's really happening is back pressure.]

(3) Add a continuous listener to `PersistedIndex`, and when a new value comes in do something like:

```go
s.pendingLock.Lock()
batchCount := 0
while batchCount < len(s.pendingLock) && s.pending[batchCount].index < newPersistedIndex {
	s.pending[batchCount].ACK()
	batchCount++
}
s.pending = s.pending[batchCount:]
s.pendingLock.Unlock()
```

We can jump in a zoom to talk through this if you want, I know the pipeline logic is a little arcane and this is getting long.

Reply:
Actually, one more small change that would help with freeing up memory: instead of saving the original `Batch` structure, just save the `ACK` callback. Fingers crossed, but that might be enough to keep pending events from hanging around in memory once we make the appropriate queue changes.

Reply:
> Well, simplicity is nice but I think limiting the entire shipper queue to at most one batch is enough of a constraint that it's worth tracking some metadata.

I think I agree that having a simple reference client is nice, but we also need the Beats client to be performant to ensure users experience use of the shipper as a strict improvement over what we had before. Tiago has started integrating the shipper in the V2 inputs repository and since that is a from scratch system it may be better suited as the reference client.

I think we need both follow @faec's suggestion above to make the listening async and also  make the change described in https://github.com/elastic/elastic-agent-shipper/issues/97 to ensure we do not double buffer events (once in beats, once in the shipper).

I don't know that we need to do both of these things in this PR however, it is fine as a starting point. Certainly the queue changes can be separate.

Reply:
The queue changes are definitely separate. Making this code async could be an in-place revision or a second pass, but should definitely happen before we give this output to customers. (But since we're pre-beta I would be willing to approve this one if there's a followup issue.)

Reply:
I'll make it async in a follow up PR, @faec do you think we can merge the current state then? I'm still waiting for an approval :)

Reply:
Well, I wanted an explicit issue before approving, so I wrote one 😜 https://github.com/elastic/elastic-agent-shipper/issues/98 I will do one last pass of the rest of this PR